genai_factory/src/genai_factory/workflows/workflow_server.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2023 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from genai_factory import logger
+from genai_factory.config import WorkflowServerConfig
+from genai_factory.controller_client import ControllerClient
+from genai_factory.schemas import WorkflowType
+from genai_factory.sessions import SessionStore
+from genai_factory.workflows import Workflow
+
+
+class WorkflowServer:
+    def __init__(self, config: WorkflowServerConfig = None):
+        self._config = config or WorkflowServerConfig()
+        self._controller_client = None
+        self._session_store = SessionStore(self._config)
+        self._workflows: dict[str, Workflow] = {}
+
+    @property
+    def config(self) -> WorkflowServerConfig:
+        return self._config
+
+    @property
+    def controller_client(self) -> ControllerClient:
+        if not self._controller_client:
+            self._controller_client = ControllerClient(
+                controller_url=self._config.controller_url,
+                project_name=self._config.project_name,
+                username=self._config.controller_username,
+            )
+        return self._controller_client
+
+    def set_config(self, config: WorkflowServerConfig):
+        self._config = config
+        self._session_store = SessionStore(self._config)
+        for workflow in self._workflows.values():
+            workflow._server = None
```

Comment:
This is in case someone already called `_build` and then changed the configuration so the workflows need to be rebuilt as well... Let us simply say that once you `deploy` (which calls `_build` that is protected anyway) you can't change configuration.

```suggestion
        for workflow in self._workflows.values():
            workflow._server = None
```

But, what you did miss here, is that you need to reinitialize the `controller_client` with the new configuration.

Reply:
The commit suggestion is the same as the code before. Anyway if I understand what you meant, I added the controller initialization like this:
```python
    def set_config(self, config: WorkflowServerConfig):
        self._config = config
        self._controller_client()
        self._session_store = SessionStore(self._config)
        for workflow in self._workflows.values():
            workflow._server = None
```